### PR TITLE
[quant][pyper] Add embedding_bag weight quantize and dequantize ops

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41293 [quant][pyper] Add embedding_bag weight quantize and dequantize ops**

Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22506700](https://our.internmc.facebook.com/intern/diff/D22506700)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  int64_t embedding_cols = weight.size(1);
+
+  Tensor weight_contig = weight.contiguous(weight.suggest_memory_format());
+
+  const auto weight_data = weight.data_ptr<float>();
+  constexpr int BIT_RATE = 4;
+  constexpr int NUM_ELEM_PER_BYTE = 8 / BIT_RATE;
+  TORCH_CHECK(
+      weight_contig.size(weight.dim() - 1) % NUM_ELEM_PER_BYTE == 0,
+      "FloatToFused4BitRowwiseQuantizedOp only works for the number of "
```
Comment (radkris-git) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:65]: I think this is a nit.

File: test/quantization/test_quantized_op.py [hunk 2]
```diff
@@ -2703,23 +2703,77 @@ def test_qlinear_unpack(self, W, use_channelwise):
...
+        from caffe2.python import core, workspace
+        conversion_op = "FloatToFused8BitRowwiseQuantized"
+        if bit_rate == 4:
+            conversion_op = "FloatToFused4BitRowwiseQuantized"
+
+        def get_c2_weights(weights):
+            workspace.ResetWorkspace()
+
+            workspace.FeedBlob("weights", weights)
+            workspace.RunOperatorOnce(
```
Comment (radkris-git) [inline test/quantization/test_quantized_op.py:2724]: Using C2 operators gives confidence that PT ops work. Do you plan to keep the comparison with C2 for the longer term as well?

Comment (supriyar) [reply test/quantization/test_quantized_op.py:2724]: No, we will add reference implementations eventually once C2 is deprecated.

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp (hunk 1 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:65]: You mean this check isn't required?

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          (embedding_cols + NUM_ELEM_PER_BYTE - 1) / NUM_ELEM_PER_BYTE +
+          2 * sizeof(at::Half))};
+  auto output = at::empty(
+      output_shape,
+      weight_contig.options().dtype(at::kByte),
+      weight_contig.suggest_memory_format());
+  auto* output_data = output.data_ptr<uint8_t>();
+  const auto output_columns = output.size(output.dim() - 1);
+
+  for (int row = 0; row < embedding_rows; ++row) {
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:86]: does performance matter, or is this a reference implementation?  Could probably parallelize if needed (same for the other op)

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+          0,
+          std::min<int>(
+              lrintf((X - Xmin) * inverse_scale), (1 << BIT_RATE) - 1));
+      if (col % NUM_ELEM_PER_BYTE == 0) {
+        output_row[col / NUM_ELEM_PER_BYTE] = quantized;
+      } else {
+        output_row[col / NUM_ELEM_PER_BYTE] |=
+            (quantized << ((col % NUM_ELEM_PER_BYTE) * BIT_RATE));
+      }
+    }
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:126]: optional nit: `} // embedding_cols` for readability?

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  }
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:127]: optional nit: `} // embedding_rows` for readability?

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      weight_contig.options().dtype(at::kByte),
+      weight_contig.suggest_memory_format());
+  auto* output_data = output.data_ptr<uint8_t>();
+  const auto output_columns = output.size(output.dim() - 1);
+
+  for (int row = 0; row < embedding_rows; ++row) {
+    const float* input_row = weight_data + row * embedding_cols;
+    std::uint8_t* output_row = output_data + row * output_columns;
+
+    at::Half* output_row_scale_zp = reinterpret_cast<at::Half*>(
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:90]: optional readability nit: if this is packed at the end of a row, maybe we can move the code down to be below the weight packing, so the code structure follows the data format?

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    output_row_scale_zp[0] = scale;
+    output_row_scale_zp[1] = Xmin;
+
+    for (int col = 0; col < embedding_cols; ++col) {
+      float X = input_row[col];
+      std::uint8_t quantized = std::max(
+          0,
+          std::min<int>(
+              lrintf((X - Xmin) * inverse_scale), (1 << BIT_RATE) - 1));
+      if (col % NUM_ELEM_PER_BYTE == 0) {
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:120]: maybe we can add a quick comment on the data format? i.e. index 0 is packed in the low bits and index 1 in the high bits, etc

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_unpack.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+#include <torch/library.h>
+
+namespace at {
+namespace native {
+namespace {
+
+Tensor qembeddingbag_byte_unpack(const Tensor& packed_weight) {
+  const auto input_rows = packed_weight.size(0);
+  const auto input_columns = packed_weight.size(1);
+  int output_columns = input_columns - 2 * sizeof(float);
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qembeddingbag_unpack.cpp:11]: nit: add comment that the 2 is coming from scale+zp?

**Review by vkuzo — COMMENTED** [review]
lg, feel free to ignore the comments if this implementation will be replaced in the near future

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp (hunk 3 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:86]: This op is a temporary solution until we de-couple quantize and packing. We can re-visit optimizations then.

Comment (vkuzo) [reply aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:86]: makes sense

**Review by vkuzo — APPROVED** [review]

File: aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp (hunk 1 above)
Comment (radkris-git) [reply aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:65]: The error message says "FloatToFused4BitRowwiseQuantizedOp". :) It should be "qembeddingbag_4bit_prepack only works for the number of columns a multiple of 2".

**Review by radkris-git — APPROVED** [review]

**Pull request closed without merging**